drop `index.shard.check_on_startup: fix`

vladimirdolzhenko opened the pull request:
* drop `index.shard.check_on_startup: fix` setting
* put a corrupt marker if it is detected during start up check that index

Relates #31389

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2558,6 +2567,104 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+        final IndexMetaData indexMetaData = IndexMetaData.builder(indexShard.indexSettings().getIndexMetaData())
+            .settings(Settings.builder()
+                .put(indexShard.indexSettings.getSettings())
+                .put(IndexSettings.INDEX_CHECK_ON_STARTUP.getKey(), "checksum"))
```
Does this test also work if we set this to `true`? If so, could we choose randomly between these two settings, and rename the test appropriately?

DaveCTurner replied:
It sounds to me like it's doing the wrong thing here, but these things are subtle. @ywelsch might know: if `check_on_startup: true` finds a corruption it doesn't currently mark the shard as corrupted. I think it should. WDYT?

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2558,6 +2567,104 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+        // start shard with checksum - it has to pass successfully
+
+        final ShardRouting shardRouting = ShardRoutingHelper.initWithSameId(indexShard.routingEntry(),
+            primary ? RecoverySource.StoreRecoverySource.EXISTING_STORE_INSTANCE : RecoverySource.PeerRecoverySource.INSTANCE
```
`primary` is always true here.

vladimirdolzhenko replied:
++

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2558,6 +2567,104 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+        assertThat("store is clean",
+            corruptedMarkerCount.get(), equalTo(0));
+
+        // storeProvider that does not perform check index on close - it is corrupted
```
I don't fully understand this comment, or what this `storeProvider` is for. The heart of this test seems to be:

    expectThrows(IndexShardRecoveryException.class, () -> newStartedShard(p -> corruptedShard, primary));

We also want the corruption marker to exist as soon as this is thrown, not after the shard is subsequently closed. If we need to test that opening a shard with a corruption marker fails, and that the corruption marker isn't removed when it's closed, then I think that should be a separate test.

vladimirdolzhenko replied:
Indeed, that heart of the test is 
`expectThrows(IndexShardRecoveryException.class, () -> newStartedShard(p -> corruptedShard, primary));`

the problem with default store provider - that it checks index on close. In case of corrupted index - it fails on close (as it's corrupted) - and we have to close it to avoid resource leakage.

therefore the key point of adjusted storeProvider is 
```
baseDirectoryWrapper.setCheckIndexOnClose(false);
```

Agree to split it into 2 tests - the 2nd one with `opening a shard with a corruption marker fails`

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2558,6 +2567,104 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+            Files.walk(indexPath)
+                .filter(p -> Files.isRegularFile(p) && IndexWriter.WRITE_LOCK_NAME.equals(p.getFileName().toString()) == false)
+                .toArray(Path[]::new);
+        CorruptionUtils.corruptFile(random(), filesToCorrupt);
```
Does this _definitely_ corrupt a file that will be checked? It doesn't, for instance, hit a translog file in a generation that's early enough that it's no longer needed?

vladimirdolzhenko replied:
It indeed corrupts only index file due to 
`final Path indexPath = shardPath.getDataPath().resolve("index");`

translog files are stored in another place - `shardPath.getDataPath().resolve("translog");`

vladimirdolzhenko commented:
@elasticmachine test this please

vladimirdolzhenko commented:
@DaveCTurner I addressed your comments - tried ran test twice but it fails on completely irrelevant part (passed on private CI) - could you please have another look into this PR ?

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2558,6 +2564,148 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+                .put(IndexSettings.INDEX_CHECK_ON_STARTUP.getKey(), randomFrom("true", "checksum")))
+            .build();
+
+        final Path indexPath = corruptIndexFile(shardPath);
```
I think this test should create the corruption marker directly, rather than relying on the check-on-startup behaviour. Essentially, it should fail to start even in the presence of a corruption marker, even if the index itself is not corrupt.

vladimirdolzhenko replied:
agree

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2558,6 +2564,148 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+        };
+        Files.walkFileTree(indexPath, corruptedVisitor);
+
+        assertThat("store is clean",
```
I think I'd prefer statements to go all on one line if possible. There's a few places where that could be done, such as this one.

DaveCTurner commented:
I think this should go into 6.x too - your labels indicate this will only be merged into 7.0. We also need a PR to update the breaking changes docs for 6.x to note that `fix` is no longer supported. I'm not sure if we should note that in the 6.0 breaking changes, because it's not worked since then.

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2571,6 +2577,142 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+            null, null, indexShard.engineFactory,
+            indexShard.getGlobalCheckpointSyncer(), EMPTY_EVENT_LISTENER);
+
+        expectThrows(IndexShardRecoveryException.class, () -> newStartedShard(p -> corruptedShard, true));
```
Please could you assert something about the thrown exception - ideally that its message describes the affected shard and contains an appropriate message?

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2571,6 +2577,142 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+        indexShard.flush(new FlushRequest());
+        closeShards(indexShard);
+
+        // start shard with checksum - it has to pass successfully
```
This comment seems misleading: we don't start the shard here, and we expect it not to pass successfully when we do.

vladimirdolzhenko replied:
makes sense, addressed your comments

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2571,6 +2577,142 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+
+        final Path indexPath = corruptIndexFile(shardPath);
+
+        // check that corrupt marker is *NOT* there
```
I don't think this comment is necessary here. The check is the assertion a few lines further down, and it should be clear what that assertion is doing. I think the message in the assertion could use the phrase "corruption marker" rather than "clean" for consistency.

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2571,6 +2577,142 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+        assertThat(exception1.getCause().getMessage(), equalTo(corruptionMessage + " (resource=preexisting_corruption)"));
+        closeShards(corruptedShard);
+
+        // check that corrupt marker is there
```
Similarly - this comment isn't needed here.

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -2571,6 +2577,142 @@ public void testReadSnapshotConcurrently() throws IOException, InterruptedExcept
+
+        // create corrupted marker
+        final String corruptionMessage = "fake ioexception";
+        try(Store store = createStore(indexShard.indexSettings(), shardPath)){
```
Nit: whitespace

DaveCTurner commented:
Also please update the labels on this to indicate how it will be backported. I think it should go into 6.5 too.

vladimirdolzhenko commented:
it has to be backported to 6.5 together with https://github.com/elastic/elasticsearch/pull/32281

DaveCTurner commented:
We discussed back-porting in another channel and agreed:

- this change can be made independently of #32281 
- for ease of review this PR should be made backportable to 6.5 with a small follow-up to actually remove the option in 7.0.

DaveCTurner commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -299,6 +299,10 @@ public IndexShard(
         this.checkIndexOnStartup = indexSettings.getValue(IndexSettings.INDEX_CHECK_ON_STARTUP);
+        if ("fix".equals(checkIndexOnStartup)) {
+            deprecationLogger.deprecated("Setting [index.shard.check_on_startup] is set to deprecated value [fix], "
+                + "which will be unsupported in future");
```
Suggest

> which has no effect and will not be accepted in future

vladimirdolzhenko replied:
++ thanks

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardDeprecatedSettingTests.java:
```diff
@@ -0,0 +1,54 @@
+        return false;
+    }
+
+    public void testCheckOnStartupDeprecatedValue() throws Exception {
```
I think this could be simplified to:

```
    public void testCheckOnStartupDeprecatedValue() throws Exception {
        final IndexShard newShard = newShard(true, Settings.builder().put(IndexSettings.INDEX_CHECK_ON_STARTUP.getKey(), "fix").build());
        assertWarnings("Setting [index.shard.check_on_startup] is set to deprecated value [fix], which will be unsupported in future");
        closeShards(newShard);
    }
```

(and using `assertWarnings` means it doesn't need to be in its own test fixture too)

vladimirdolzhenko replied:
already figured that out ;)

DaveCTurner commented on server/src/test/java/org/elasticsearch/index/shard/IndexShardDeprecatedSettingTests.java:
```diff
@@ -0,0 +1,54 @@
+    public void testCheckOnStartupDeprecatedValue() throws Exception {
+        final Settings settings = Settings.builder().put(IndexSettings.INDEX_CHECK_ON_STARTUP.getKey(), "fix").build();
+
+        try(ThreadContext threadContext = new ThreadContext(Settings.EMPTY)) {
```
Nit: missing whitespace after the `try`.

vladimirdolzhenko commented:
already figured that out ;)

vladimirdolzhenko commented:
@elasticmachine test this please

vladimirdolzhenko commented:
thanks @DaveCTurner 👍
